src/cirrus/documentation_utils.py

```diff
@@ -0,0 +1,288 @@
+    except KeyError:
+        LOGGER.error('Did not find a complete [doc] section in cirrus.conf'
+                     '\nSee below for an example:'
+                     '\n[doc]'
+                     '\n;sphinx_makefile_dir = /path/to/sphinx')
+        sys.exit(1)
+
+    cmd = 'cd {} && make clean html'.format(docs_root)
+
+    if make_opts:
+        # additional args were passed after --docs. Pass these to make
+        cmd = 'cd {} && make {}'.format(docs_root, ' '.join(make_opts))
+
+    local('. ./{}/bin/activate && {}'.format(venv_name, cmd))
+    LOGGER.info('Build command was "{}"'.format(cmd))
+
+
+def build_doc_artifact():
+    """
+    build sphinx documentation and create an artifact to be uploaded to
+    a remote server
+
+    Requires the following cirrus.conf section:
+    [doc]
+    sphinx_makefile_dir = /path/to/makefile
+    sphinx_doc_dir = /path/to/_build/docdir
+    artifact_dir = /path/to/archive/dir
+    """
+    LOGGER.info("Building doc archive...")
+
+    config = load_configuration()
+    artifact_name = doc_artifact_name(config)
+    arcname = os.path.basename(artifact_name).rsplit('.', 2)[0]
+    doc_dir = config['doc']['sphinx_doc_dir']
+
+    if not os.path.exists(doc_dir):
+        msg = "Documentation path: {0} Not Found".format(doc_dir)
+        LOGGER.error(msg)
+        raise RuntimeError(msg)
+
+    with tarfile.open(artifact_name, "w:gz") as tar:
+        tar.add(doc_dir, arcname=arcname, recursive=False)
+
+    LOGGER.info("Documentation artifact created at: {}".format(artifact_name))
+
+    return artifact_name
+
+
+def upload_documentation(opts):
+    """
+    Upload Sphinx documentation to a remote server.
+
+    If a tarfile exists for the current program version, that file will
+    be used. Otherwise, the documentation will be built and packaged
+    before uploading.
+
+    The argparse.Namespace that calls this functions should have an
+    option named '--docs' which takes a list of arguments which specify
+    how the documentation should be built. If no arguments are provided,
+    that indicates the documentation artifact already exists for the
+    release version being uploaded or that the default arguments -
+    'clean html' - should be used to create the documentation.
+
+    Requires the following cirrus.conf section:
+    [doc]
+    upload_method = [file_server | jenkins]
+
+    :param argparse.Namspace opts: A Namespace of upload options
+    """
+    LOGGER.info("Preparing to upload documentation...")
+    config = load_configuration()
+    doc_params = config.get('doc', {})
+
+    doc_artifact = doc_artifact_name(config)
+    if not os.path.exists(doc_artifact):
+        msg = 'Documentation tarball not found at {}'.format(doc_artifact)
+        LOGGER.error(msg)
+        raise RuntimeError(msg)
+
+    # make sure there is a valid upload method
+    if not doc_params.get('upload_method') \
+            or doc_params.get('upload_method') not in ['file_server', 'jenkins']:
```

Comment:
I'd suggest combining somewhat complex logic like this into a new method `is_valid_upload_method`